pingcap/tidb
# config, sysvar: move configs tmp-storage-* to instance sysvars

### What problem does this PR solve?


Issue Number: ref #34960 

Problem Summary:

Before, if we want to change the temporary storage path/quota in a node/instance, we have to change the configuration and restart the node.

### What is changed and how it works?

In this PR we introduce two instance scope variables `tmpdir` and `tidb_tmp_storage_quota`, and the corresponding config items in `[instance]` section.

### Check List

Tests 

- [x] Unit test (existing UT)

#### Manual Test

##### Data preparation

Setup `t1` and `t2`:

```bash
#!/bin/bash
mysql -u root -h 127.0.0.1 -D test -P 4000 -e "DROP TABLE IF EXISTS t1, t2"
mysql -u root -h 127.0.0.1 -D test -P 4000 -e "CREATE TABLE t1 (c INT); CREATE TABLE t2 (c VARCHAR(100));"
for ((i=1; i<=9999; i++)); do
mysql -u root -h 127.0.0.1 -D test -P 4000 -e "INSERT INTO t1 VALUES ($i); INSERT INTO t2 VALUES ($i);"
done
```

##### Test

There are several functions using `tmpdir`:

1. `util/disk/tempDir.go: InitializeTempDir()` would be called each time tmpdir is modified, and create a `_dir.lock` file. We expect to see this file after we modify tmpdir.

	You can just switch different `tmpdir` like `TestSwitchMultipleTmpDir`, and check that both `$tmpdir/record/` and `$tmpdir/_dir.lock` exist.

2. `executor/explain.go: getHeapProfile()` would create a temp file named like `$tmpdir/record/heapGC2006-01-02T15:04:05Z07:001`. We can use the sql `explain analyze select ...` to trigger a long-time query and modify [instance].tmpdir in another session. We expect that the modification is successful and the origin heap file names are recorded in `memoryDebugModeHandler.infoField` and shown on logs.
  	* connect to session 1 and execute:
  	```sql
	set tidb_memory_debug_mode_min_heap_inuse = 1024;
	set tidb_memory_debug_mode_alarm_ratio = 100;
	-- The following query costs about 3 seconds to execute.
	explain analyze select t1.c as c1, t2.c as c2 from t1 join t2 on t1.c = t2.c - t1.c order by c2 desc;
	```
	* connect to session 2 and execute `set global tmpdir = 'another_tmpdir';`
	* waiting the `explain analyze` in session 1 to finish. The log would show a line containing `["Memory Debug Mode"] [sql=finished]`. In this line you could see `heap profile` following the `tmpdir`. In most cases it has been set to `another_tmpdir`, but it is still reasonable if it keeps the origin value.

3. `util/misc.go: LoadTLSCertificates()` creates a key file and certificate in tmpdir. We can set `auto-tls` to `true` in `config.toml` and use the sql `alter instance reload tls` to call `LoadTLSCertificates()`. By checking the log, we expect that the paths of cert and key file keep the same as the original.
	* create a `config.toml` as below and start the tidb:
	```toml
	[security]
	auto-tls = true
	```
	* connect to tidb and check the ssl key and cert by `select @@ssl_key, @@ssl_cert;`
	* execute `alter instance reload tls;`, and the log `["execute reload tls"]` shows the ssl_key and ssl_path
	* `select @@ssl_key, @@ssl_cert;` and we expect to get the origin result

4. `util/chunk/disk.go: initDiskFile()` uses `disk.TmpDirMutex` to prevent possible conflicts since it has several statements reading tmpdir. We can perform a large query with `sort` or `join` operator under limited memory to call this method. Since the block of mutex, the modification of tmpdir would wait for the query finished.
	* do the query in session 1. Two temporary files `chunk.ListInDisk-xxxxx` and `chunk.ListInDiskOffset-xxxxx` exists on tmpdir:
	```sql
	set @@tidb_mem_quota_query = 100 << 10; -- 100KB
	set global tidb_mem_oom_action = 'log';
	-- The following query costs hundreds of seconds to execute.
	select t1.c as c1, t2.c as c2 from t1 join t2 on t1.c = t2.c - t1.c order by c2 desc;
	```
	* when the query in session 1 is going on, run `set global tmpdir = "another_tmpdir";` in session 2
	* when the query finished in session 1, the origin 2 temp `chunk` files are deleted, and `select @@global.tmpdir;` returns `another_tmpdir` in session 1.

5. `type memoryUsageAlarm struct` binds to each domain, and owns a `tmpDir` field for dumping information about OOM. We can modify `tidb_memory_usage_alarm_ratio` to trigger the alarm dumping.
	* create a `config.toml` as below and start the tidb:
	```toml
	[performance]
	# you can set some other suitable values to trigger the memory alarm
	server-memory-quota = 1048576 # 1MB
	memory-usage-alarm-ratio = 0.01
	```
	* query `select t1.c as c1, t2.c as c2 from t1 join t2 on t1.c = t2.c - t1.c order by c2 desc;`

6. `ddl/ingest/env.go: genLightningDataDir()` is called each time a ddl is starting. We expect to see a directory named `$tmpdir/tmp_ddl-<port>` after we modify the tmpdir and do another ddl.

Side effects


Documentation


### Release note



Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
1. Merge configs `tmp-storage-path` and `temp-dir` into one config `[instance].tmpdir`, and create a new corresponding instance system variable `tmpdir`
2. Move config `tmp-storage-quota` to `[instance].tidb_tmp_storage_quota`, and create a new corresponding instance system variable `tidb_tmp_storage_quota`
```
